File: ports/laravel/tests/Conformance/VectorsTest.php
<?php

namespace I18nKeyless\Laravel\Tests\Conformance;

use I18nKeyless\Laravel\ApiClient;
use I18nKeyless\Laravel\KeylessTranslator;
use I18nKeyless\Laravel\Locale;
use I18nKeyless\Laravel\Miss;
use I18nKeyless\Laravel\Tests\TestCase;
use Illuminate\Http\Client\ConnectionException;
use Illuminate\Http\Client\Factory;
use Illuminate\Http\Client\Request;
use Illuminate\Support\Facades\Cache;
use Illuminate\Support\Facades\Http;
use Illuminate\Support\Facades\Log;
use Illuminate\Support\Sleep;

/**
 * Replays the shared conformance vectors of the monorepo (conformance/vectors/*.json).
 *
 * Not replayed, on purpose: replace.json (placeholders are Laravel's `:name`),
 * storage-keys.json and unique-id.json cases (a device storage contract; a server
 * sends no id), the usage-*.json device and custom-handler cases, the
 * forceTemporary / originLanguage / custom-handler cases (not exposed by this
 * package), and the queue id rule (this package dedupes by key AND context).
 */
final class VectorsTest extends TestCase
{
    private const VECTORS = __DIR__.'/../../../../conformance/vectors';

    /** @return array<string, mixed> */
    private function vector(string $name): array
    {
        $path = self::VECTORS."/{$name}.json";
        if (! is_file($path)) {
            $this->markTestSkipped("no conformance vector at {$path}");
        }

        return json_decode((string) file_get_contents($path), true, flags: JSON_THROW_ON_ERROR);
    }

    public function test_resolve_lang(): void
    {
        foreach ($this->vector('resolve-lang')['cases'] as $case) {
            $input = $case['input'];
            $this->assertSame(
                $case['expected'],
                Locale::resolve($input['tag'], $input['supported'] ?? null, $input['fallback'] ?? null),
                $case['name']
            );
        }
    }

    public function test_languages(): void
    {
        foreach ($this->vector('languages')['cases'] as $case) {
            switch ($case['check']) {
                case 'availableLangs':
                    $this->assertSame($case['expected'], Locale::AVAILABLE_LANGS, $case['name']);
                    break;
                case 'rename':
                    // No v2 dialect in this package: the v2 code is simply not a language.
                    $this->assertNull(Locale::toLang($case['input']), $case['name']);
                    $this->assertTrue(Locale::isLang($case['expected']), $case['name']);
                    break;
                case 'stillAvailable':
                    foreach ($case['input'] as $code) {
                        $this->assertTrue(Locale::isLang($code), $case['name']);
                    }
                    break;
                case 'absent':
                    $this->assertFalse(Locale::isLang($case['input']), $case['name']);
                    break;
                case 'regionalized':
                    $regionalized = array_values(array_filter(Locale::AVAILABLE_LANGS, fn ($code) => str_contains($code, '-')));
                    sort($regionalized);
                    $this->assertSame($case['expected'], $regionalized, $case['name']);
                    break;
                default:
                    $this->fail("unknown check {$case['check']}");
            }
        }
    }

    public function test_app_store_locales(): void
    {
        $vector = $this->vector('app-store-locales');
        foreach ($vector['cases'] as $case) {
            $this->assertSame($case['expected'], Locale::toAppStoreLocale($case['input']));
        }
        $this->assertCount($vector['distinctSlots'], array_unique(Locale::APP_STORE_LOCALES));
    }

    public function test_storage_key(): void
    {
        foreach ($this->vector('storage-key')['cases'] as $case) {
            $this->assertSame($case['expected'], Miss::lookupKeyFor($case['input']['key'], $case['input']['context'] ?? null), $case['name']);
        }
    }

    public function test_namespace_resolution(): void
    {
        foreach ($this->vector('namespace')['cases'] as $case) {
            if ($case['fn'] !== 'resolveNamespace') {
                continue; // originLanguage (user generated content) is not exposed by this package
            }
            $options = $case['input']['options'] ?? [];
            $this->assertSame(
                $case['expected'],
                KeylessTranslator::resolveNamespace($options['namespace'] ?? null, $case['input']['config']['defaultNamespace'] ?? null),
                $case['name']
            );
        }
    }

    public function test_retry_decision(): void
    {
        foreach ($this->vector('retry-decision')['cases'] as $case) {
            $decision = ApiClient::decide($case['input']['status'], $case['input']['statusText']);
            $this->assertSame($case['expected']['action'], $decision['action'], json_encode($case['input']));
            if (isset($case['expected']['error'])) {
                $this->assertSame($case['expected']['error'], $decision['error'], json_encode($case['input']));
            }
        }
    }

    public function test_backoff_schedule(): void
    {
        $vector = $this->vector('backoff');
        $client = new ApiClient(new Factory, 'k', 'https://api.test', $vector['timeoutMs'] / 1000, $vector['delaysMs']);
        $this->assertSame($vector['maxAttempts'], $client->maxAttempts());
        $this->assertSame($vector['timeoutMs'] / 1000, config('i18n-keyless.timeout'));
        $this->assertSame($vector['delaysMs'], config('i18n-keyless.retry'));
        foreach ($vector['cases'] as $case) {
            $this->assertSame($case['expected']['waitMs'], $client->delayAfter($case['input']['failedAttempt']), $case['name']);
        }
    }

    public function test_backoff_scenarios(): void
    {
        $vector = $this->vector('backoff');
        foreach ($vector['scenarios'] as $scenario) {
            Sleep::fake();
            $attempts = 0;
            $responses = $scenario['responses'];
            $http = new Factory;
            $http->fake(['https://api.test/*' => function () use (&$attempts, &$responses, $vector) {
                $attempts++;
                $answer = array_shift($responses);
                if (isset($answer['networkError'])) {
                    throw new ConnectionException($answer['networkError']);
                }
                if (! empty($answer['timeout'])) {
                    throw new ConnectionException('cURL error 28: Operation timed out after '.$vector['timeoutMs'].' milliseconds');
                }
                if (! empty($answer['invalidJson'])) {
                    return Http::response('{not json', 200);
                }

                return Http::response(isset($answer['body']) ? json_encode($answer['body']) : '', $answer['status']);
            }]);
            $client = new ApiClient($http, 'k', 'https://api.test', 10, $vector['delaysMs']);

            $result = $client->fetchDictionary('en', 'default', $scenario['name'] === '304 ends the call at once' ? 'W/"x"' : null);

            $this->assertSame($scenario['expected']['attempts'], $attempts, $scenario['name']);
            Sleep::assertSequence(array_map(fn (int $ms) => Sleep::for($ms)->milliseconds(), $scenario['expected']['sleepsMs']));
            $expected = $scenario['expected']['result'];
            $this->assertSame($expected['ok'], $result['ok'], $scenario['name']);
            if (isset($expected['notModified'])) {
                $this->assertSame($expected['notModified'], $result['notModified'], $scenario['name']);
            }
            if (isset($expected['error']) && $expected['error'] !== 'HTTP 403') {
                // Guzzle always fills a reason phrase, so an empty status text cannot be replayed on the wire:
                // that case is covered by test_retry_decision through ApiClient::decide().
                $this->assertSame($expected['error'], $result['error'], $scenario['name']);
            }
        }
    }

    public function test_dictionary_request(): void
    {
        foreach ($this->vector('dictionary-request')['cases'] as $case) {
            if (isset($case['expected']['handler'])) {
                continue; // custom handlers are not exposed by this package
            }
            $input = $case['input'];
            $http = new Factory;
            $http->fake(['*' => Http::response(['ok' => true, 'data' => ['translations' => []], 'error' => '', 'message' => ''])]);
            $client = new ApiClient($http, $input['config']['API_KEY'], $input['config']['API_URL'] ?? ApiClient::DEFAULT_URL);
            $client->fetchDictionary($input['targetLanguage'], $input['namespace'] ?? 'default', $input['knownEtag'] ?? null, $input['lastRefresh']);

            $expected = $case['expected'];
            $http->assertSent(function (Request $request) use ($expected) {
                $this->assertSame($expected['url'], $request->url());
                $this->assertSame($expected['method'], $request->method());
                foreach (['Content-Type', 'Authorization', 'If-None-Match'] as $header) {
                    if (isset($expected['headers'][$header])) {
                        $this->assertSame($expected['headers'][$header], $request->header($header)[0] ?? null);
                    }
                }
                $this->assertSame(ApiClient::VERSION, $request->header('Version')[0] ?? null);
                // A Laravel application is always a server: the sdk header says so, and no device id travels.
                $this->assertSame(ApiClient::SDK, $request->header('sdk')[0] ?? null);
                $this->assertFalse($request->hasHeader('unique_id'));
                $this->assertFalse($request->hasHeader('If-None-Match') && ! isset($expected['headers']['If-None-Match']));

                return true;
            });
        }
    }

    public function test_dictionary_response(): void
    {
        Log::spy();
        $warnings = [];
        foreach ($this->vector('dictionary-response')['cases'] as $case) {
            if (isset($case['expected']['warning'])) {
                $warnings[] = 'i18n-keyless: '.$case['expected']['warning'];
            }
            Sleep::fake();
            $attempts = 0;
            $responses = $case['responses'] ?? [$case['response']];
            $http = new Factory;
            $http->fake(['*' => function () use (&$attempts, &$responses) {
                $attempts++;
                $answer = array_shift($responses);

                return Http::response(isset($answer['body']) ? json_encode($answer['body']) : '', $answer['status'], $answer['headers'] ?? []);
            }]);
            $client = new ApiClient($http, $case['input']['config']['API_KEY'], ApiClient::DEFAULT_URL);
            $knownEtag = $case['input']['knownEtag'] ?? null;

            $result = $client->fetchDictionary($case['input']['targetLanguage'], 'default', $knownEtag);

            $expected = $case['expected'];
            if (isset($expected['attempts'])) {
                $this->assertSame($expected['attempts'], $attempts, $case['name']);
            }
            if ($expected['result'] === null) {
                $this->assertTrue(! $result['ok'] || $result['notModified'], $case['name']);
                $this->assertSame([], $result['translations'], $case['name']);
            } else {
                $this->assertTrue($result['ok'], $case['name']);
                $this->assertSame($expected['result']['data']['translations'], $result['translations'], $case['name']);
                $this->assertSame($expected['result']['etag'] ?? null, $result['etag'], $case['name']);
            }
            // The ETag a caller keeps afterwards: the new one, else the one it already had on a 304.
            $remembered = $result['notModified'] ? $knownEtag : $result['etag'];
            $this->assertSame($expected['etagRemembered'], $remembered, $case['name']);
            $this->assertSame(
                $expected['nextRequest']['url'],
                $client->dictionaryUrl($case['input']['targetLanguage'], 'default', $remembered, '1700000000'),
                $case['name']
            );
        }
        foreach ($warnings as $warning) {
            Log::shouldHaveReceived('warning')->with($warning)->once();
        }
    }

    public function test_translate_request(): void
    {
        foreach ($this->vector('translate-request')['cases'] as $case) {
            $options = $case['input']['options'] ?? [];
            if (isset($case['expected']['handler']) || isset($options['forceTemporary']) || isset($options['originLanguage'])) {
                continue; // not exposed by this package
            }
            $config = $case['input']['config'];
            $http = new Factory;
            $http->fake(['*' => Http::response(['ok' => true, 'data' => ['translation' => []], 'error' => '', 'message' => ''])]);
            $client = new ApiClient($http, $config['API_KEY'], $config['API_URL'] ?? ApiClient::DEFAULT_URL);
            $miss = new Miss(
                $case['input']['key'],
                $options['context'] ?? null,
                KeylessTranslator::resolveNamespace($options['namespace'] ?? null, $config['defaultNamespace'] ?? null),
            );

            $client->translate([$miss], $config['languages']['primary'], $config['languages']['supported']);

            $expected = $case['expected'];
            $http->assertSent(function (Request $request) use ($expected) {
                $this->assertSame($expected['url'], $request->url());
                $this->assertSame($expected['method'], $request->method());
                $this->assertSame($expected['headers']['Content-Type'], $request->header('Content-Type')[0] ?? null);
                $this->assertSame($expected['headers']['Authorization'], $request->header('Authorization')[0] ?? null);
                $this->assertSame(ApiClient::VERSION, $request->header('Version')[0] ?? null);
                $this->assertSame(ApiClient::SDK, $request->header('sdk')[0] ?? null);
                $this->assertFalse($request->hasHeader('unique_id'));
                $this->assertSame($expected['body'], $request->data());

                return true;
            });
        }
    }

    public function test_queue_scenarios(): void
    {
        foreach ($this->vector('queue')['scenarios'] as $scenario) {
            if (! is_array($scenario['calls'])) {
                continue; // "31 distinct keys": the in-flight peak is a promise-level property of the SDK queue
            }
            $skip = false;
            foreach ($scenario['calls'] as $call) {
                $options = $call['options'] ?? [];
                if (isset($options['originLanguage']) || isset($options['forceTemporary']) || isset($options['context'])) {
                    $skip = true; // not exposed, or deduplicated differently (this package keeps one POST per context)
                }
            }
            if ($skip) {
                continue;
            }
            $this->refreshApplication();
            Http::fake([
                'https://api.test/translate/*' => Http::response(['ok' => true, 'data' => ['translations' => $scenario['translations'] ?? []], 'error' => '', 'message' => '']),
                'https://api.test/translate' => Http::response(['ok' => true, 'data' => ['translation' => []], 'error' => '', 'message' => '']),
            ]);
            $this->app->setLocale('en');

            foreach ($scenario['calls'] as $call) {
                i18nk($call['key'], [], null, null, $call['options']['namespace'] ?? null);
            }
            $this->app->terminate();

            $posts = Http::recorded(fn (Request $r) => $r->method() === 'POST' && $r->url() === 'https://api.test/translate')->count();
            $this->assertSame($scenario['expected']['requests'], $posts, $scenario['name']);
        }
    }

    public function test_translation_lookup(): void
    {
        foreach ($this->vector('translation-lookup')['cases'] as $case) {
            $options = $case['input']['options'] ?? [];
            if (isset($options['originLanguage']) || isset($options['forceTemporary']) || isset($options['replace']) || isset($options['unpersistedNamespace'])) {
                continue; // not exposed by this package (placeholders are Laravel's `:name`)
            }
            $this->refreshApplication();
            $store = $case['input']['store'];
            $this->app['config']->set('i18n-keyless.primary', $store['primary']);
            $this->app['config']->set('i18n-keyless.namespace', $store['defaultNamespace'] ?? 'default');
            Http::fake(['https://api.test/translate/*' => Http::response(['ok' => true, 'data' => ['translations' => []], 'error' => '', 'message' => ''])]);
            $namespace = KeylessTranslator::resolveNamespace($options['namespace'] ?? null, $store['defaultNamespace'] ?? null);
            Cache::forever(
                'i18n-keyless:'.substr(sha1('test-key'), 0, 8).":dict:{$namespace}:{$store['currentLanguage']}",
                ['translations' => $store['translations'], 'etag' => null, 'fetched_at' => time(), 'failed' => false]
            );
            $this->app->setLocale($store['currentLanguage']);

            $text = i18nk($case['input']['key'], [], $options['context'] ?? null, null, $options['namespace'] ?? null);

            $this->assertSame($case['expected']['text'], $text, $case['name']);
            $queued = array_values(array_unique(array_map(
                fn (Miss $miss) => $miss->namespace,
                $this->app->make(KeylessTranslator::class)->pendingMisses()
            )));
            $this->assertSame(array_map(fn ($q) => $q['namespace'], $case['expected']['queued']), $queued, $case['name']);
        }
    }

    public function test_usage_reporting_matches_the_node_runtime(): void
    {
        $vector = $this->vector('usage-reporting');
        $node = null;
        foreach ($vector['cases'] as $case) {
            if (($case['input']['package'] ?? null) === 'node') {
                $node = $case['expected'];
            }
        }
        $this->assertNotNull($node);
        // `laravel` is registered on the API as a server label with the `node` rules.
        $this->assertSame('node', $node['runtime']);
        $this->assertSame('laravel', ApiClient::SDK);
        $this->assertTrue($node['recordsUsage']);
        $this->assertTrue($node['sendsUsage']);
        $this->assertFalse($node['sendsUniqueId']);
        $this->assertTrue(config('i18n-keyless.usage'));
        $this->assertSame(10, \I18nKeyless\Laravel\DictionaryStore::USAGE_FLUSH_SECONDS);

        // End to end: a served key is recorded, and the map leaves after the response.
        Http::fake([
            'https://api.test/translate/last-used-translations' => Http::response(['ok' => true, 'message' => '']),
            'https://api.test/translate/*' => Http::response(['ok' => true, 'data' => ['translations' => ['Bonjour' => 'Hello']], 'error' => '', 'message' => '']),
        ]);
        $this->app->setLocale('en');
        $this->assertSame('Hello', __('Bonjour'));
        $this->app->terminate();
        Http::assertSent(fn (Request $r) => $r->url() === 'https://api.test/translate/last-used-translations'
            && ! $r->hasHeader('unique_id')
            && $r->hasHeader('sdk', 'laravel')
            && $r->data()['translationsUsageByNamespace'] === ['default' => ['Bonjour' => gmdate('Y-m-d')]]);
    }

    public function test_usage_request(): void
    {
        foreach ($this->vector('usage-request')['cases'] as $case) {
            $config = $case['input']['config'];
            if (isset($case['expected']['handler'])) {
                continue; // custom handlers are not exposed by this package
            }
            $http = new Factory;
            $http->fake(['*' => Http::response(['ok' => true, 'message' => ''])]);
            $client = new ApiClient($http, $config['API_KEY'], $config['API_URL'] ?? ApiClient::DEFAULT_URL);

            $result = $client->sendUsage($config['languages']['primary'], $case['input']['usage']);

            $expected = $case['expected'];
            if (($expected['http'] ?? true) === false) {
                $this->assertFalse($result['sent'], $case['name']);
                $http->assertNothingSent();
                continue;
            }
            $this->assertTrue($result['ok'], $case['name']);
            $http->assertSent(function (Request $request) use ($expected) {
                $this->assertSame($expected['url'], $request->url());
                $this->assertSame($expected['method'], $request->method());
                $this->assertSame($expected['headers']['Content-Type'], $request->header('Content-Type')[0] ?? null);
                $this->assertSame($expected['headers']['Authorization'], $request->header('Authorization')[0] ?? null);
                $this->assertSame(ApiClient::VERSION, $request->header('Version')[0] ?? null);
                $this->assertSame(ApiClient::SDK, $request->header('sdk')[0] ?? null);
                $this->assertFalse($request->hasHeader('unique_id'));
                $this->assertSame($expected['body'], $request->data());

                return true;
            });
        }
    }

    public function test_a_server_sends_no_device_id(): void
    {
        $vector = $this->vector('unique-id');
        $this->assertStringContainsString('A server runtime sends no id', $vector['description']);
        $http = new Factory;
        $http->fake(['*' => Http::response(['ok' => true, 'data' => ['translations' => []], 'error' => '', 'message' => ''])]);
        (new ApiClient($http, 'k', 'https://api.test'))->fetchDictionary('en', 'default', null);
        $http->assertSent(fn (Request $r) => ! $r->hasHeader('unique_id') && $r->hasHeader('sdk', 'laravel'));
    }
}

File: ports/rails/test/conformance/vectors_test.rb
# frozen_string_literal: true

require "test_helper"

# Replays the shared conformance vectors of the monorepo (conformance/vectors/*.json).
#
# Not replayed, on purpose: replace.json (placeholders are I18n's `%{name}`),
# storage-keys.json and unique-id.json cases (a device storage contract; a server
# sends no id), the usage-*.json device and custom-handler cases, the
# forceTemporary / originLanguage / custom-handler cases (not exposed by this
# gem), and the queue id rule (this gem dedupes by key AND context).
class VectorsTest < I18nKeylessTest::Case
  # assert_equal with a nil expectation is deprecated in minitest; the vectors use null for "no value".
  def assert_value(expected, actual, message = nil)
    expected.nil? ? assert_nil(actual, message) : assert_equal(expected, actual, message)
  end

  def vector(name)
    path = File.join(I18nKeylessTest::VECTORS, "#{name}.json")
    skip "no conformance vector at #{path}" unless File.file?(path)
    JSON.parse(File.read(path))
  end

  def client(api_key = "k", api_url = "https://api.test", timeout: 10, delays: [500, 1500])
    api = I18nKeyless::ApiClient.new(api_key: api_key, api_url: api_url, timeout: timeout, retry_delays: delays,
                                     logger: Logger.new(log))
    api.sleeper = ->(ms) { sleeps << ms }
    api
  end

  def test_resolve_lang
    vector("resolve-lang")["cases"].each do |c|
      input = c["input"]
      assert_value c["expected"], I18nKeyless::Locale.resolve(input["tag"], supported: input["supported"], fallback: input["fallback"]), c["name"]
    end
  end

  def test_languages
    vector("languages")["cases"].each do |c|
      case c["check"]
      when "availableLangs"
        assert_equal c["expected"], I18nKeyless::Locale::AVAILABLE_LANGS, c["name"]
      when "rename"
        # No v2 dialect in this gem: the v2 code is simply not a language.
        assert_nil I18nKeyless::Locale.to_lang(c["input"]), c["name"]
        assert I18nKeyless::Locale.lang?(c["expected"]), c["name"]
      when "stillAvailable"
        c["input"].each { |code| assert I18nKeyless::Locale.lang?(code), c["name"] }
      when "absent"
        refute I18nKeyless::Locale.lang?(c["input"]), c["name"]
      when "regionalized"
        assert_equal c["expected"], I18nKeyless::Locale::AVAILABLE_LANGS.select { |code| code.include?("-") }.sort, c["name"]
      else
        flunk "unknown check #{c['check']}"
      end
    end
  end

  def test_app_store_locales
    v = vector("app-store-locales")
    v["cases"].each { |c| assert_equal c["expected"], I18nKeyless::Locale.to_app_store_locale(c["input"]) }
    assert_equal v["distinctSlots"], I18nKeyless::Locale::APP_STORE_LOCALES.values.uniq.length
  end

  def test_storage_key
    vector("storage-key")["cases"].each do |c|
      assert_equal c["expected"], I18nKeyless::Miss.lookup_key_for(c["input"]["key"], c["input"]["context"]), c["name"]
    end
  end

  def test_namespace_resolution
    vector("namespace")["cases"].each do |c|
      next unless c["fn"] == "resolveNamespace" # originLanguage (user generated content) is not exposed by this gem

      options = c["input"]["options"] || {}
      assert_equal c["expected"],
                   I18nKeyless::Translator.resolve_namespace(options["namespace"], c["input"].dig("config", "defaultNamespace")),
                   c["name"]
    end
  end

  def test_retry_decision
    vector("retry-decision")["cases"].each do |c|
      decision = I18nKeyless::ApiClient.decide(c["input"]["status"], c["input"]["statusText"])
      assert_equal c["expected"]["action"], decision.action, c["input"].to_json
      assert_equal c["expected"]["error"], decision.error, c["input"].to_json if c["expected"].key?("error")
    end
  end

  def test_backoff_schedule
    v = vector("backoff")
    api = client(timeout: v["timeoutMs"] / 1000, delays: v["delaysMs"])
    assert_equal v["maxAttempts"], api.max_attempts
    assert_equal v["timeoutMs"] / 1000, I18nKeyless::Config.new({}).timeout
    assert_equal v["delaysMs"], I18nKeyless::Config.new({}).retry
    v["cases"].each do |c|
      assert_value c["expected"]["waitMs"], api.delay_after(c["input"]["failedAttempt"]), c["name"]
    end
  end

  def test_backoff_scenarios
    v = vector("backoff")
    v["scenarios"].each do |scenario|
      WebMock.reset!
      sleeps.clear
      attempts = 0
      responses = scenario["responses"].dup
      stub_request(:get, %r{https://api\.test/.*}).to_return do |_request|
        attempts += 1
        answer = responses.shift
        raise SocketError, answer["networkError"] if answer.key?("networkError")
        raise Net::OpenTimeout, "execution expired" if answer["timeout"]

        body = answer["invalidJson"] ? "{not json" : (answer.key?("body") ? answer["body"].to_json : "")
        { status: [answer["status"], answer["statusText"] || ""], body: body }
      end
      api = client(delays: v["delaysMs"])

      result = api.fetch_dictionary("en", "default", scenario["name"] == "304 ends the call at once" ? 'W/"x"' : nil)

      assert_equal scenario["expected"]["attempts"], attempts, scenario["name"]
      assert_equal scenario["expected"]["sleepsMs"], sleeps, scenario["name"]
      expected = scenario["expected"]["result"]
      assert_equal expected["ok"], result.ok, scenario["name"]
      assert_equal expected["notModified"], result.not_modified, scenario["name"] if expected.key?("notModified")
      assert_equal expected["error"], result.error, scenario["name"] if expected.key?("error")
    end
  end

  def test_dictionary_request
    vector("dictionary-request")["cases"].each do |c|
      next if c["expected"].key?("handler") # custom handlers are not exposed by this gem

      WebMock.reset!
      input = c["input"]
      stub_request(:get, %r{.*}).to_return(status: 200, body: envelope({}).to_json)
      api = client(input["config"]["API_KEY"], input["config"]["API_URL"] || I18nKeyless::ApiClient::DEFAULT_URL)
      api.fetch_dictionary(input["targetLanguage"], input["namespace"] || "default", input["knownEtag"], input["lastRefresh"])

      expected = c["expected"]
      assert_requested(:get, expected["url"], times: 1) do |request|
        %w[Content-Type Authorization If-None-Match].each do |header|
          assert_equal expected["headers"][header], request.headers[header], "#{c['name']}: #{header}" if expected["headers"].key?(header)
        end
        assert_equal I18nKeyless::ApiClient::VERSION, request.headers["Version"], c["name"]
        # A Rails application is always a server: the sdk header says so, and no device id travels.
        assert_equal I18nKeyless::ApiClient::SDK, request.headers["Sdk"], c["name"]
        refute request.headers.key?("Unique-Id"), c["name"]
        refute request.headers.key?("If-None-Match") && !expected["headers"].key?("If-None-Match"), c["name"]
        true
      end
    end
  end

  def test_dictionary_response
    warnings = []
    vector("dictionary-response")["cases"].each do |c|
      warnings << "i18n-keyless: #{c['expected']['warning']}" if c["expected"].key?("warning")
      WebMock.reset!
      sleeps.clear
      attempts = 0
      responses = c["responses"] || [c["response"]]
      stub_request(:get, %r{.*}).to_return do |_request|
        attempts += 1
        answer = responses.shift
        { status: [answer["status"], answer["statusText"] || ""], body: answer.key?("body") ? answer["body"].to_json : "",
          headers: answer["headers"] || {} }
      end
      api = client(c["input"]["config"]["API_KEY"], I18nKeyless::ApiClient::DEFAULT_URL)
      known_etag = c["input"]["knownEtag"]

      result = api.fetch_dictionary(c["input"]["targetLanguage"], "default", known_etag)

      expected = c["expected"]
      assert_equal expected["attempts"], attempts, c["name"] if expected.key?("attempts")
      if expected["result"].nil?
        assert !result.ok || result.not_modified, c["name"]
        assert_equal({}, result.translations, c["name"])
      else
        assert result.ok, c["name"]
        assert_equal expected["result"]["data"]["translations"], result.translations, c["name"]
        assert_value expected["result"]["etag"], result.etag, c["name"]
      end
      # The ETag a caller keeps afterwards: the new one, else the one it already had on a 304.
      remembered = result.not_modified ? known_etag : result.etag
      assert_value expected["etagRemembered"], remembered, c["name"]
      assert_equal expected["nextRequest"]["url"],
                   api.dictionary_url(c["input"]["targetLanguage"], "default", remembered, "1700000000"), c["name"]
    end
    warnings.each { |warning| assert_equal 1, log.string.scan(warning).length, warning }
  end

  def test_translate_request
    vector("translate-request")["cases"].each do |c|
      options = c["input"]["options"] || {}
      next if c["expected"].key?("handler") || options.key?("forceTemporary") || options.key?("originLanguage")

      WebMock.reset!
      config = c["input"]["config"]
      stub_request(:post, %r{.*}).to_return(status: 200, body: { ok: true, data: { translation: {} }, error: "", message: "" }.to_json)
      api = client(config["API_KEY"], config["API_URL"] || I18nKeyless::ApiClient::DEFAULT_URL)
      miss = I18nKeyless::Miss.new(c["input"]["key"], options["context"],
                                   I18nKeyless::Translator.resolve_namespace(options["namespace"], config["defaultNamespace"]))

      api.translate([miss], config["languages"]["primary"], config["languages"]["supported"])

      expected = c["expected"]
      assert_requested(:post, expected["url"], times: 1) do |request|
        assert_equal expected["headers"]["Content-Type"], request.headers["Content-Type"], c["name"]
        assert_equal expected["headers"]["Authorization"], request.headers["Authorization"], c["name"]
        assert_equal I18nKeyless::ApiClient::VERSION, request.headers["Version"], c["name"]
        assert_equal I18nKeyless::ApiClient::SDK, request.headers["Sdk"], c["name"]
        refute request.headers.key?("Unique-Id"), c["name"]
        assert_equal expected["body"], JSON.parse(request.body), c["name"]
        true
      end
    end
  end

  def test_queue_scenarios
    vector("queue")["scenarios"].each do |scenario|
      next unless scenario["calls"].is_a?(Array) # "31 distinct keys": the in-flight peak is a promise-level property of the SDK queue

      skip_scenario = scenario["calls"].any? do |call|
        options = call["options"] || {}
        options.key?("originLanguage") || options.key?("forceTemporary") || options.key?("context")
      end
      next if skip_scenario # not exposed, or deduplicated differently (this gem keeps one POST per context)

      WebMock.reset!
      configure
      stub_request(:get, %r{https://api\.test/translate/.*}).to_return(status: 200, body: envelope(scenario["translations"] || {}).to_json)
      stub_translate
      I18n.locale = :en

      scenario["calls"].each do |call|
        I18nKeyless.translate(call["key"], namespace: call.dig("options", "namespace"))
      end
      flush

      assert_requested(:post, "#{I18nKeylessTest::API}/translate", times: scenario["expected"]["requests"]) if scenario["expected"]["requests"].positive?
      assert_not_requested(:post, "#{I18nKeylessTest::API}/translate") if scenario["expected"]["requests"].zero?
    end
  end

  def test_translation_lookup
    vector("translation-lookup")["cases"].each do |c|
      options = c["input"]["options"] || {}
      next if %w[originLanguage forceTemporary replace unpersistedNamespace].any? { |o| options.key?(o) } # placeholders are I18n's `%{name}`

      WebMock.reset!
      store = c["input"]["store"]
      configure(primary: store["primary"], namespace: store["defaultNamespace"])
      stub_request(:get, %r{https://api\.test/translate/.*}).to_return(status: 200, body: envelope({}).to_json)
      namespace = I18nKeyless::Translator.resolve_namespace(options["namespace"], store["defaultNamespace"])
      seed_dictionary(store["currentLanguage"], store["translations"], namespace: namespace)
      I18n.locale = store["currentLanguage"]

      text = I18nKeyless.translate(c["input"]["key"], context: options["context"], namespace: options["namespace"])

      assert_equal c["expected"]["text"], text, c["name"]
      queued = I18nKeyless.translator.pending_misses.map(&:namespace).uniq
      assert_equal c["expected"]["queued"].map { |q| q["namespace"] }, queued, c["name"]
    end
  end

  def test_usage_reporting_matches_the_node_runtime
    v = vector("usage-reporting")
    node = v["cases"].find { |c| c.dig("input", "package") == "node" }["expected"]
    # `rails` is registered on the API as a server label with the `node` rules.
    assert_equal "node", node["runtime"]
    assert_equal "rails", I18nKeyless::ApiClient::SDK
    assert node["recordsUsage"]
    assert node["sendsUsage"]
    refute node["sendsUniqueId"]
    rails_label = v["serverLabels"]["cases"].find { |c| c["label"] == "rails" }
    assert rails_label, "usage-reporting.json must list `rails` as a server label"
    assert rails_label["expected"]
    assert I18nKeyless::Config.new({}).usage?
    assert_equal 10, I18nKeyless::DictionaryStore::USAGE_FLUSH_SECONDS

    # End to end: a served key is recorded, and the map leaves after the response.
    stub_dictionary("en", { "Bonjour" => "Hello" })
    stub_usage
    I18n.locale = :en
    assert_equal "Hello", I18n.t("Bonjour")
    flush
    assert_requested(:post, "#{I18nKeylessTest::API}/translate/last-used-translations") do |request|
      refute request.headers.key?("Unique-Id")
      assert_equal "rails", request.headers["Sdk"]
      assert_equal({ "default" => { "Bonjour" => today } }, JSON.parse(request.body)["translationsUsageByNamespace"])
      true
    end
  end

  def test_usage_request
    vector("usage-request")["cases"].each do |c|
      config = c["input"]["config"]
      next if c["expected"].key?("handler") # custom handlers are not exposed by this gem

      WebMock.reset!
      stub_request(:post, %r{.*}).to_return(status: 200, body: { ok: true, error: "", message: "" }.to_json)
      api = client(config["API_KEY"], config["API_URL"] || I18nKeyless::ApiClient::DEFAULT_URL)

      result = api.send_usage(config["languages"]["primary"], c["input"]["usage"])

      expected = c["expected"]
      if expected["http"] == false
        refute result.sent, c["name"]
        assert_not_requested(:post, %r{.*})
        next
      end
      assert result.ok, c["name"]
      assert_requested(:post, expected["url"], times: 1) do |request|
        assert_equal expected["headers"]["Content-Type"], request.headers["Content-Type"], c["name"]
        assert_equal expected["headers"]["Authorization"], request.headers["Authorization"], c["name"]
        assert_equal I18nKeyless::ApiClient::VERSION, request.headers["Version"], c["name"]
        assert_equal I18nKeyless::ApiClient::SDK, request.headers["Sdk"], c["name"]
        refute request.headers.key?("Unique-Id"), c["name"]
        assert_equal expected["body"], JSON.parse(request.body), c["name"]
        true
      end
    end
  end

  def test_a_server_sends_no_device_id
    v = vector("unique-id")
    assert_includes v["description"], "A server runtime sends no id"
    stub_request(:get, %r{.*}).to_return(status: 200, body: envelope({}).to_json)
    client.fetch_dictionary("en", "default", nil)
    assert_requested(:get, "https://api.test/translate/en?last_refresh=") do |request|
      !request.headers.key?("Unique-Id") && request.headers["Sdk"] == "rails"
    end
  end
end
